### ISSUE TITLE
Transactions with 'db.useTransaction' when multiple (same) connections are around.

### ISSUE THREAD

**Issue by AxelG1** [vincentlauvlwj/Ktorm#65]
Hello.

I had a situation in which I had multiple `Database` s with the same connection parameters around. Rollback had no effect when I used the transaction scope as a method `db.useTransaction { ... }`, but succeeded whenever I used the function `useTransaction`, i.e. on the global instance.

I stepped through the code. `rollback` SQL is called reliably but in the case described above has no effect. 

As soon as I reduced the multiple connections to one, both ways of employing the transaction scope worked.

Maybe having multiple same connections is against the concept? Then of course it can also be correct undefined behaviour.

Great work by the way!

**Comment by vincentlauvlwj** [comment]
How can I reproduce that? Can you show me your code?

**Comment by AxelG1** [comment]
I condensed it to this:

```
// inside DBUtils:
fun ktormConnect(config: JsonObject): Database {
    val host = config.getString("host")
    val port = config.getInteger("port")
    val db = config.getString("db")
    val u = config.getString("user")
    val p = config.getString("pass")
    val dbUrl = "jdbc:mysql://%s:%d/%s".format(host, port, db)
    val driver = "com.mysql.cj.jdbc.Driver"
    val dataBase = Database.connect(dbUrl, driver, u, p)
    return dataBase
}
...
// then:
val dbConfig = config.getJsonObject("db")
val db = DbUtils.ktormConnect(dbConfig)
val db2 = DbUtils.ktormConnect(dbConfig)    // this line makes the difference, if it's active rollback has no effect
...
// somewhere else I use a function copied from transaction tests:
fun testTransaction() {
    class DummyException : Exception()

    try {
        //useTransaction {
        //Database.global.useTransaction {
        db.useTransaction {
            PendingConfirmations.insert {
                it.token to "administration"
                it.type to 23
                it.email to "Hong Kong"
                it.time to TimeUtils.shared.now()
            }
            assert(PendingConfirmations.count() == 2)
            throw DummyException()
        }
    } catch (e: DummyException) {
        assert(PendingConfirmations.count() == 1)
    }
}
```
I'm still slow formating code here...

**Comment by vincentlauvlwj** [comment]
Thank you for your feedback. I reproduce this bug with the following code: 

```kotlin
fun main() {
    val db = Database.connect("jdbc:mysql://127.0.0.1:3306/ktorm", driver = "com.mysql.jdbc.Driver", user = "root")
    val db2 = Database.connect("jdbc:mysql://127.0.0.1:3306/ktorm", driver = "com.mysql.jdbc.Driver", user = "root")

    class DummyException : Exception()

    try {
        db.useTransaction {
            Departments.insert {
                it.name to "test"
                it.location to "New York"
            }

            println(Departments.count())
            throw DummyException()
        }
    } catch (e: DummyException) {
        println(Departments.count())
    }
}
```

**Comment by vincentlauvlwj** [comment]
Related issues are: #47, #41, #27

**Comment by vincentlauvlwj** [comment]
This bug is caused by the design of `Database.global`, which seems to be a mistake now. 

After we call `connect` function, Ktorm will save the latest created database instance to `Database.global` automatically, then obtain it when executing a SQL. 

`Departments.insert`, `Departments.count` or any other functions always use the global instance `Database.global` to execute their generated SQL. That's why the transaction doesn't work here, we are opening an transaction with `db`, but executing our SQL commands with `db2`.

```kotlin
fun main() {
    val db = Database.connect("jdbc:mysql://127.0.0.1:3306/ktorm", driver = "com.mysql.jdbc.Driver", user = "root")
    val db2 = Database.connect("jdbc:mysql://127.0.0.1:3306/ktorm", driver = "com.mysql.jdbc.Driver", user = "root")

    class DummyException : Exception()

    try {
        db.useTransaction {                     // open transaction with db
            Departments.insert {                // insert a record with db2 (Database.global)
                it.name to "test"
                it.location to "New York"
            }

            println(Departments.count())        // query the record count with db2 (Database.global)
            throw DummyException()
        }
    } catch (e: DummyException) {
        println(Departments.count())            // query the record count with db2 (Database.global)
    }
}
```

**Comment by vincentlauvlwj** [comment]
I noticed the problem of `Database.global` months ago. I've been working on deprecating it for long and it's just finished now. I will release it as Ktorm version 2.7  after updating the documents. 

There is a workaround before 2.7 released. `db { }` can change the value `Database.global` in the scope of the closure. You can use it temporally: 

```kotlin
fun main() {
    val db = Database.connect("jdbc:mysql://127.0.0.1:3306/ktorm", driver = "com.mysql.jdbc.Driver", user = "root")
    val db2 = Database.connect("jdbc:mysql://127.0.0.1:3306/ktorm", driver = "com.mysql.jdbc.Driver", user = "root")

    class DummyException : Exception()

    db {                                        // Database.global will change to db in the scope of the closure
        try {
            db.useTransaction {                 // open transaction with db
                Departments.insert {            // insert a record with db2 (Database.global)
                    it.name to "test"
                    it.location to "New York"
                }

                println(Departments.count())    // query the record count with db2 (Database.global)
                throw DummyException()
            }
        } catch (e: DummyException) {
            println(Departments.count())        // query the record count with db2 (Database.global)
        }
    }
}
```

**Comment by AxelG1** [comment]
Thank you!
For now I changed my connection management, but I'm looking forward to the next release. Cheers!

**Comment by vincentlauvlwj** [comment]
This issue is fixed in Ktorm 2.7, see https://ktorm.liuwj.me/en/about-deprecating-database-global.html
